### PR TITLE
Ampers - Abinnet + Lily - Scrabble

### PR DESCRIPTION
**Pull request by lillers1122**
# Scrabble
Congratulations! You're submitting your assignment.

## Comprehension Questions

Feature | Feedback
--- | ---
What are the advantages of using git when collaboratively working on one code base? | Git helps facilitate collaboration on one code base by making it possible for all collaborators to work on their own computers and forces version control for the overall progress.
Why do you think we chose to make the `score` method in the `Scoring` class a **class method** instead of an **instance method**? | Because you want to be able to compare scores across different instances. Also centralizing scoring in one class is more efficient because we do not over-process data when we call on it from other classes.
Describe an example of a test you wrote for a _nominal case_. | Player test: it 'returns the score of the word' do / player_1 = Scrabble::Player.new("lily") / player_1.play("ada").must_equal 4 /end / is a nominal test case because it tests that a method works the way it should for a reasonable input. Describe an example of a test you wrote for an _edge case_. | Scoring test:   it 'returns nil for empty words' do / Scrabble::Scoring.score('').must_be_nil / end / tests an edge case test because '' is not a expected input. 
Does your code use any methods from the `Enumerable` mixin? If so, where and why was it helpful? |    In our tile_bag.rb file we used the enumerable ".sum" in this method: /def tiles_remaining / @tile_bag.values.sum / end / It was helpful because it made our code shorter and more efficient. And, it was easier to troubleshoot.
What was one method you and your pair used to debug code? | We used awesome_print a lot. We also white boarded before we started.
Is there a specific piece of code you'd like feedback on? | We chose to put our letter and their values as arrays in the scoring.rb file. Should we have used hashes? We weren't sure which to do. Also, we got a little bogged down by how our classes communicate with each other and use each others' methods.
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? | 1. Communication style established at start; also assumed positive intent with all feedback 2. Coded together with the assumption we were not pursing the BEST path, but were nevertheless pursing valid paths.

### CODE REVIEW

File: lib/scoring.rb [hunk 1]
```diff
@@ -1,9 +1,69 @@
...
+        else
+          @word_total_points = nil
+          return @word_total_points
+        end
+      end
+      return @word_total_points
     end
 
     def self.highest_score_from(array_of_words)
+      max_points = 0
```
Comment (CheezItMan) [inline lib/scoring.rb]: This method breaks when given two words, both with the same score but the 2nd one has 7 letters.  Instead of returning the 7-letter-word it returns the 1st word in case of a tie, every time.

File: specs/scoring_spec.rb [hunk 2]
```diff
@@ -42,21 +42,32 @@
...
     it 'returns the only word in a length-1 array' do
+      Scrabble::Scoring.highest_score_from(["word"]).must_equal "word"
     end
 
     it 'returns the highest word if there are two words' do
+      Scrabble::Scoring.highest_score_from(["pig", "cat"]).must_equal "pig"
     end
 
     it 'if tied, prefer a word with 7 letters' do
+      Scrabble::Scoring.highest_score_from(["aaaaaad", "qqqqqj"]).must_equal "aaaaaad"
```
Comment (CheezItMan) [inline specs/scoring_spec.rb]: This only tests that the 7-letter word wins, if it comes first in the array.

To test it adequately you'd need to also test:
```ruby
Scrabble::Scoring.highest_score_from(["qqqqqj", "aaaaaad"]).must_equal "aaaaaad"
```

File: specs/scoring_spec.rb [hunk 3]
```diff
@@ -42,21 +42,32 @@
...
     end
 
     it 'if tied and no word has 7 letters, prefers the word with fewer letters' do
+      Scrabble::Scoring.highest_score_from(["qqqqqq", "zxd"]).must_equal "qqqqqq"
```
Comment (CheezItMan) [inline specs/scoring_spec.rb]: Look at the description here!  The word with **more** letters is winning!

Like above you should also scramble the order.

File: specs/scoring_spec.rb [hunk 4]
```diff
@@ -42,21 +42,32 @@
...
+
     end
 
     it 'returns the first word of a tie with same letter count' do
+      Scrabble::Scoring.highest_score_from(["adbf", "egch"]).must_equal "adbf"
```
Comment (CheezItMan) [inline specs/scoring_spec.rb:76]: Again swapping around the order is needed.

File: lib/scoring.rb [hunk 5]
```diff
@@ -1,9 +1,69 @@
+require 'awesome_print'
+
 module Scrabble
   class Scoring
+    attr_reader :word_total_points
+
     def self.score(word)
+      @word_array = word.upcase.split("")
```
Comment (CheezItMan) [inline lib/scoring.rb:8]: This works, but can you think of a more efficient data structure to hold the point values?

File: lib/tile_bag.rb [hunk 6]
```diff
@@ -0,0 +1,101 @@
+require 'pry'
+require_relative 'scoring.rb'
+require 'awesome_print'
+
+module Scrabble
+  class TileBag
+    attr_reader :tile_bag
+
+    def initialize
+      @tile_bag =
```
Comment (CheezItMan) [inline lib/tile_bag.rb:10]: There's a problem here.  The way you have it, each letter has an equal chance of being drawn, but there are 9 "A" tiles, while only 2 "P".

File: lib/tile_bag.rb [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+      #   "Q" => 10, "Z" => 10
+      # }
+
+    end
+
+    def draw_tiles(num)
+      drawn_tiles = []
+
+      num.times do
+      drawn_tiles.push(@tile_bag.keys.sample)
```
Comment (CheezItMan) [inline lib/tile_bag.rb:82]: Works, but the odds are off.  Also what if all the "P" tiles are drawn, you'll end up with a **negative** number of tiles!

File: lib/player.rb [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+  class Player
+    attr_accessor :name
+    attr_reader :plays, :tiles_hand
+
+    def initialize(name)
+      @name = name
+      @plays = []
+      @score = 0
+      @won = false
+      @hsw = "ERROR"
```
Comment (CheezItMan) [inline lib/player.rb:16]: ???

File: lib/player.rb [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+      return @won
+    end
+
+    def highest_scoring_word
+      @hsw = Scrabble::Scoring.highest_score_from(@plays)
+      return @hsw
+    end
+
+    def highest_word_score
+      score = Scrabble::Scoring.score(@hsw)
```
Comment (CheezItMan) [inline lib/player.rb:54]: This method **assumes** that `highest_scoring_word` has **already** been called.  What if it hasn't?  

You should instead `score = Scrabble::Scoring.score(self.highest_scoring_word)`

File: specs/player_spec.rb [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+
+# Get that nice colorized output
+Minitest::Reporters.use! Minitest::Reporters::SpecReporter.new
+
+describe 'Player' do
+  describe 'initialize' do
+    it 'can initialize new player' do
+      player = Scrabble::Player.new("Abinnet")
+      player.must_be_instance_of Scrabble::Player
+      player.name.must_equal "Abinnet"
```
Comment (CheezItMan) [inline specs/player_spec.rb:15]: You should also check that `plays` is an empty array.

File: specs/player_spec.rb [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+    it 'returns the score of the word' do
+      player_1 = Scrabble::Player.new("lily")
+      player_1.play("ada").must_equal 4
+    end
+  end
+
+  describe 'total_score' do
+    it 'it return the total score' do
+      player_1 = Scrabble::Player.new("Abbinet")
+      player_1.play("ada")
```
Comment (CheezItMan) [inline specs/player_spec.rb:46]: You should also test for what happens if they've played more words than required to win.  The  words after they score 100 should be omitted from the total score.

File: specs/player_spec.rb [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+      player_1.won?.must_equal false
+    end
+  end
+
+  describe 'highest_scoring_word' do
+    it 'returns the highest scoring word' do
+      player_1 = Scrabble::Player.new("Abbinet")
+      player_1.play("ada")
+      player_1.play("dog")
+      player_1.highest_scoring_word.must_equal "DOG"
```
Comment (CheezItMan) [inline specs/player_spec.rb:71]: You should also test it with no words, and with too many words.

File: specs/player_spec.rb [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+    end
+  end
+
+  describe 'highest_word_score' do
+    it 'returns the highest score' do
+      player_1 = Scrabble::Player.new("Abbinet")
```
Comment (CheezItMan) [inline specs/player_spec.rb:77]: Again similar tests to the above.

File: specs/player_spec.rb [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+    end
+  end
+
+  describe 'tiles' do
+    it 'returns collection of tiles in player hand' do
+      game_1 = Scrabble::TileBag.new
+      player_1 = Scrabble::Player.new("Isabel")
+      player_1.tiles.must_equal []
+      player_1.draw_tiles(game_1)
+      player_1.tiles_hand.length.must_equal 7
```
Comment (CheezItMan) [inline specs/player_spec.rb:90]: You should also have some way to test that the correct letters were removed from the tilebag and placed in the player's hand.

File: specs/player_spec.rb [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+    end
+  end
+
+  describe 'draw_tiles' do
+    it 'can fill hand to 7 tiles' do
+      game_1 = Scrabble::TileBag.new
```
Comment (CheezItMan) [inline specs/player_spec.rb:96]: This test hasn't changed from above.

File: specs/tile_bag_spec.rb [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+require_relative '../lib/tile_bag'
+
+# Get that nice colorized output
+Minitest::Reporters.use! Minitest::Reporters::SpecReporter.new
+
+describe 'TileBag' do
+  describe 'initialize' do
+    it 'correctly initializes a bag of tiles' do
+      game_1 = Scrabble::TileBag.new
+      game_1.tile_bag["C"].must_equal 2
```
Comment (CheezItMan) [inline specs/tile_bag_spec.rb:14]: More testing required.

File: specs/tile_bag_spec.rb [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+
+  describe 'draw_tiles' do
+    it 'draws random tiles' do
+      game_1 = Scrabble::TileBag.new
+      draw_1 = game_1.draw_tiles(3)
+      draw_1.length.must_equal 3
+    end
+
+    it 'removes random tiles from tile_bag' do
+      game_1 = Scrabble::TileBag.new
```
Comment (CheezItMan) [inline specs/tile_bag_spec.rb:26]: Good

File: specs/tile_bag_spec.rb [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+      end
+
+      sum.must_equal 95
+    end
+
+    # xit 'edgecase test' do
+    # end
+
+    describe 'tiles_remaining' do
+      it 'returns number of tiles remaining in bag' do
```
Comment (CheezItMan) [inline specs/tile_bag_spec.rb:42]: It would be better to set `sum_control = game_1.tiles_remaining` then draw and then compare the difference.

**Comment by CheezItMan** [comment]
# Scrabble
## What We're Looking For

Feature | Feedback
--- | ---
**General** |
Answered comprehension questions |  Check, Over-processing data, really isn't a thing, at least now how you're describing it.  I'm glad you worked together well. 
Both Teammates contributed to the codebase | Both teammate have commits!
Regular Commits with meaningful commit messages |  Check, Ok number of commits, but you really need to focus on the functionality you added in the commit messages, not vague messages about tests or waves.  
**Scoring class** |
`score` method |  Check
Uses appropriate data structure to store the letter score | You could do a more efficient job tracking letter scores.  Yours however works. 
Appropriately handles edge cases | Score handles the edge cases.
`highest_score_from` method |  **DOES NOT WORK**  You are only returning the 1st element from a tie, and your tests **always** assume the 1st element is the winner.  You did **NOT** look at the conditions for the winning word.  Read the question more carefully.  
Appropriately handles edge cases |  **NO** See above and in inline notes. 
Test implementations for `highest_score_from`|  **Some** good tests, but this method is **not** throughly tested.  See my inline notes.  
**Player class** |
Uses an array to keep track of words played |  Check
Uses existing code to create `highest_scoring_word` and `highest_word_score` |  Your `highest_word_score` only works if the prior method was called first.  That's a seriously dangerous assumption.  
Returns `true` or `false` from `won?` method |  Check
Tests edge cases on the `play(word)` |  **No** see my inline notes.  
**TileBag class** |
Uses a data structure to keep the set of default tiles in the bag | You have a data structure, but the TileBag gives each letter an **equal**  chance to be drawn even though for example vowels are present in much greater numbers.  You also have it set up to allow the `TileBag` to have  a **negative** number of a specific letter!  See my inline notes.  
`draw_tiles` method uses the parameter to draw a set of random tiles |  Check, but see the notes.
`tiles_remaining` method returns a count |  Check
Tests `draw_tiles` edge cases |  **No** edge-cases tested.  
Summary |  You both need to seriously look at your tests and make sure they are through.  If you had you would have spotted the errors in the `highest_score_from` method.  `TileBag` and `Player` bugs are harder to spot, but your testing was very thin.  Try to test all possibilities, at this stage writing too many tests is better than too few!  You hit a lot of the learning goals, except for the TDD requirement.  You need to be testing **all** the requirements.  This is something I'd like you both to focus on this week.
